### DISCUSSION TITLE
Difference between that-depends and modern-di

### DISCUSSION THREAD

**Discussion by Jonas1312** [modern-python/that-depends#116] [Q&A] [answered]
Hi,

what's the difference between that-depends and [modern-di](https://github.com/modern-python/modern-di)? 

Thanks

**Comment by lesnik512** [comment, marked as answer]
@Jonas1312 Hi!
1. The main difference: `modern-di` has scopes. Thanks to them it is easier to manage dependencies lifecycle.
2. `modern-di` does not have global state. All state lives in containers of different scopes including overrides (mocks). `That depends` has global state inside providers and local state inside contextvars
3. `modern-di` requires integration logic for web-frameworks to create container on each request and to resolve dependencies. `That depends` mostly does not require integration logic except for context resources.
4. `that-depends` is production-ready, tested in production in many projects. `modern-di` is in early state of development and I have only started to integrate it to my work project

So `modern-di` is more powerful but in not production-ready yet, `that-depends` is simpler and ready for production

**Reply by Jonas1312** [reply]
Thanks for your answer

Ok I see, I feel like modern-di is in-between that-depends and dishka, that does scope-management too

**Reply by lesnik512** [reply]
Yes, you are right) I was inspired by dishka's scopes management, but I wanted to have the dependencies graph like in dependency-injector and to have separate package for every integration to be able to update them independently. For example, litestar integration is stuck on 3.12 for now, but fastapi is working on 3.13
